File: apps/backend/src/routes/llm.ts
import { Hono } from "hono"
import { authenticate } from "../auth.js"

export const llmRouter = new Hono()

// Proxy OpenAI-compatible chat completions to OpenAI.
// The sidecar sends requests here when OPENAI_API_KEY is unavailable in the
// packaged env. The backend injects the real key server-side.
//
// authenticate is load-bearing: without it this route is an open relay that spends our
// OpenAI key for anyone who finds the URL, and bypasses metering entirely. Usage is
// charged by the caller up front via /interactions/reserve, so we do not charge here.
const DEFAULT_OPENAI_BASE = "https://api.openai.com/v1"

function aiCreditsBase(): string {
  return (process.env["OPENAI_BASE_URL"] || DEFAULT_OPENAI_BASE).replace(/\/+$/, "")
}

llmRouter.all("/proxy/*", authenticate, async (c) => {
  const apiKey = process.env["OPENAI_API_KEY"]
  if (!apiKey) return c.json({ error: "OPENAI_API_KEY not configured" }, 500)

  const upstreamPath = c.req.path.replace(/^\/api\/llm\/proxy/, "")
  const body = await c.req.text().catch(() => null)

  const headers: Record<string, string> = {
    Authorization: `Bearer ${apiKey}`,
  }

  const contentType = c.req.header("content-type")
  if (contentType) headers["Content-Type"] = contentType

  // Accept-Encoding disabled — Cloudflare Workers handle decompression
  // automatically and the upstream Content-Encoding is not forwarded.
  const res = await fetch(`${aiCreditsBase()}${upstreamPath}`, {
    method: c.req.method,
    headers,
    body: body || undefined,
  })

  if (!res.ok) {
    const text = await res.text().catch(() => "")
    return c.json(
      { error: `OpenAI upstream failed (${res.status})`, detail: text },
      res.status as 400 | 500 | 502,
    )
  }

  // Forward streaming SSE responses as-is. Use the global Response, not
  // c.newResponse, so the stream type matches across Node/Bun/Workers runtimes.
  const ct = res.headers.get("content-type") || ""
  if (ct.includes("text/event-stream") || ct.includes("application/json")) {
    return new Response(res.body, {
      status: res.status,
      headers: { "Content-Type": ct, "Cache-Control": "no-store" },
    })
  }

  return new Response(res.body, { status: res.status, headers: { "Content-Type": ct } })
})

// Legacy endpoint kept to avoid 404 for any clients still hitting it
llmRouter.post("/stream", (c) =>
  c.json({ error: "Use POST /api/llm/proxy/chat/completions instead" }, 410),
)

File: server/src/yomi/app/routes/llm.py
"""/api/llm — OpenAI-compatible proxy (apps/backend/src/routes/llm.ts).

The sidecar sends chat-completion requests here when OPENAI_API_KEY is
unavailable in the packaged env; the backend injects the real key server-side.
Authenticated — without it this route is an open relay. Usage is charged by the
caller up front via /interactions/reserve; nothing is charged here.
"""

from __future__ import annotations

import logging

import httpx
from fastapi import APIRouter, Depends, Request
from fastapi.responses import JSONResponse, Response, StreamingResponse

from yomi.app.deps import get_current_user
from yomi.conf import settings

logger = logging.getLogger(__name__)

llm_router = APIRouter(prefix="/api/llm")

DEFAULT_OPENAI_BASE = "https://api.openai.com/v1"

MIRRORED_METHODS = ["GET", "POST", "PUT", "PATCH", "DELETE", "OPTIONS", "HEAD"]


def ai_credits_base() -> str:
    return (settings.openai_base_url or DEFAULT_OPENAI_BASE).rstrip("/")


async def _proxy_handler(
    request: Request,
    upstream_path: str,
    user=Depends(get_current_user),  # noqa: B008
):
    api_key = settings.openai_api_key
    if not api_key:
        return JSONResponse({"error": "OPENAI_API_KEY not configured"}, 500)

    body = await request.body()

    headers: dict[str, str] = {"Authorization": f"Bearer {api_key}"}
    content_type = request.headers.get("content-type")
    if content_type:
        headers["Content-Type"] = content_type

    try:
        async with httpx.AsyncClient(timeout=120) as client:
            upstream = await client.request(
                request.method,
                f"{ai_credits_base()}/{upstream_path}",
                headers=headers,
                content=body or None,
            )
    except httpx.HTTPError as err:
        logger.error("[yomi/llm] upstream request failed: %s", err)
        return JSONResponse({"error": f"OpenAI upstream failed ({err})", "detail": ""}, 502)

    if upstream.status_code >= 400:
        detail = upstream.text
        return JSONResponse(
            {"error": f"OpenAI upstream failed ({upstream.status_code})", "detail": detail},
            upstream.status_code,
        )

    ct = upstream.headers.get("content-type") or ""
    response_headers = {"Content-Type": ct}
    if "text/event-stream" in ct or "application/json" in ct:
        response_headers["Cache-Control"] = "no-store"
    return StreamingResponse(
        upstream.aiter_bytes(),
        status_code=upstream.status_code,
        headers=response_headers,
    )


# Register each method separately (distinct route names → unique OpenAPI
# operation IDs); the single TS handler serves every verb.
for _method in MIRRORED_METHODS:
    llm_router.add_api_route(
        "/proxy/{upstream_path:path}",
        _proxy_handler,
        methods=[_method],
        name=f"llm_proxy_{_method.lower()}",
    )


@llm_router.post("/stream")
async def legacy_stream() -> Response:
    return JSONResponse({"error": "Use POST /api/llm/proxy/chat/completions instead"}, 410)